Prop306: Add revisions suggested in Torspec PR #87 (#98)
torproject/torspec: proposals/306-ipv6-happy-eyeballs.txt

   This section describes the necessary changes for address selection to
   implement Prop306.

2.1. Address Handling Changes

   To be able to handle Happy Eyeballs in Tor, we will need to modify the
   data structures used for connections to entry nodes, namely the extend info
   structure.

   Entry nodes are usually guards, but some clients don't use guards:

     * Bootstrapping clients can connect to fallback directory mirrors or
       authorities

     * v3 single onion services can use IPv4 or IPv6 addresses to connect
       to introduction and rendezvous points, and

     * Clients can be configured to disable entry guards

   Bridges are out of scope for this proposal, because Tor does not support
   multiple IP addresses in a single bridge line.

   The extend info structure should contain both an IPv4 and an IPv6 address.
   This will allow us to try IPv4 and the IPv6 addresses should both be
   available on a relay and the client is dual-stack.

>    When parsing relay descriptors and filling in the extend info data

Reviewer: ```suggestion
   When parsing relay descriptors, onion service descriptors, or introduce cells, and filling in the extend info data
```

Reviewer: Fixed in c935bad.

   structure, we need to fill in both the IPv4 and IPv6 address if both are
   available. If only one family is available for a relay (IPv4 or IPv6), we
   should fill in the address for preferred family and leave the alternate
   family null.

> 

Reviewer: I think we're missing a section here about bootstrapping changes. I'll commit all my suggested changes, then try writing it up.

Reviewer: Added in 2bb59d4.

2.2. Guard Selection Changes

   In Proposal 283, we have allowed microdescriptor consensus documents
   to contain IPv6 addresses. As most clients download microdescriptor
   consensuses, Prop283 makes it possible to implement Prop306.

   When we select candidates for the entry node, we should have an option
   for the number of entry guards. The parameter is as follows:

     * NumIPv6Guards NUM

   If UseEntryGuards is set to 1, we will try to pick NUM routers that support
   IPv6 for our primary guard list, which is the set of routers we strongly

>    prefer when connecting to the Tor network. If NUM is 0, we try to learn the

Reviewer: ```suggestion
   prefer when connecting to the Tor network. If NUM is auto (the default), we try to learn the
```

>    number from the guard-n-primary-guards-ipv6 consensus parameter, and default

Reviewer: Just use the same name for the consensus parameter and torrc option.
```suggestion
   number from the NumIPv6Guards consensus parameter, and default
```

Reviewer: Fixed in 99c5b9b.

>    to 1 if the consensus parameter isn’t set.

Reviewer: The initial implementation should default to 0. Once we've tested it, we can increase the default to 1.
```suggestion
   to 0 if the consensus parameter isn’t set.
```

Reviewer: I decided not to make this change.

   The default value for NumIPv6Guards should be 0.

3. Relay Connections

   If there is an existing authenticated connection, we must use it similar
   to how we used it pre-Prop306.

   If there is no existing authenticated connection for an entry node, we
   must attempt to connect using the first available, allowed, and preferred

>    address.

Reviewer: ```suggestion
   address. (Using the existing Client IPv4 and IPv6 options.)
```

Reviewer: Fixed in c935bad.

   We should also allow falling back to the alternate address. For this,
   a design will be given in Section 3.1.

3.1. TCP Connection to Preferred Address On First TCP Success

   In this design, we will connect via TCP to the first preferred address.
   On a failure or after a 250 ms delay, we attempt to connect via TCP to
   the alternate address. On a success, Tor attempts to authenticate and
   closes the other connection.

   This design is close to RFC 8305 and is similar to how Happy Eyeballs
   is implemented in a web browser.

3.2. Handling Connection Successes And Failures

   Should a connection to a entry node succeed and is authenticated via TLS,
   we can then use the connection. In this case, we should cancel all other
   connection timers and in-progress connections. Cancelling the timers is
   necessary so we don't attempt new unnecessary connections when our
   existing connection is successful, preventing denial-of-service risks.

   However, if we fail all available and allowed connections, we should tell
   the rest of Tor that the connection has failed. This is so we can attempt
   another entry node.

> 3.3. Connection Attempt Delays

Reviewer: This section is more complicated than it needs to be.

Let's have a single ConnectionAttemptDelay, which defaults to 250ms. Users can adjust it using a torrc option, or we can adjust it network-wide using a consensus parameter.

There should be limits on the ConnectionAttemptDelay:
* the RFC minimum of 10ms (you'll need to add this text)
* a maximum of 30s (as already described in the text)

Reviewer: Fixed in 7a947b9.

   As mentioned in [TEOR-P306-REP], initially, clients should prefer IPv4
   by default. The Connection Attempt Delay, or delay between IPv4 and IPv6
   connections should be 250 ms. This is to avoid the overhead from tunneled
   IPv6 connections.

>    The Minimum Connection Attempt Delay should not be dynamically adjusted

Reviewer: ```suggestion
   The Connection Attempt Delay should not be dynamically adjusted
```

Reviewer: Fixed in 7a947b9.

>    as it adds privacy risks. This value should be fixed at 10 ms as per

Reviewer: ```suggestion
   as it adds privacy risks. This value should be fixed at 250 ms as per
```

Reviewer: Fixed in 7a947b9.

>    RFC 8305 and could be adjusted using this proposed consensus parameter:

Reviewer: ```suggestion
   RFC 8305 and could be adjusted using this torrc option or consensus parameter:
```

Reviewer: Fixed in 7a947b9.

     * ConnectionAttemptDelay N msec

>    Where N is the number of milliseconds for the network-wide Minimum

Reviewer: ```suggestion
   Where N is the number of milliseconds for the network-wide
```

Reviewer: Fixed in 7a947b9.

   Connection Attempt Delay.

>    The Maximum Connection Attempt Delay should also not be dynamically adjusted

Reviewer: ```suggestion
   The Minimum and Maximum Connection Attempt Delays should also not be dynamically adjusted
```

Reviewer: Fixed in 7a947b9.

    * ClientPreferIPv6ORPort as 0 (for load-balancing reasons so we don't
      overload IPv6-only guards)

>     * ConnectionAttemptDelay 10 (for 10 msec, the minimum delay between

Reviewer: ```suggestion
    * ConnectionAttemptDelay 250 (for 250 msec, the default delay between
```

Reviewer: Fixed in 7a947b9.

      IPv4 and IPv6)

   One thing to note is that clients should be able to connect with the above
   options on IPv4-only, dual-stack, and IPv6-only networks, and they should

>    also work if ClientPreferIPv6ORPort is 1. This means we shouldn't expect

Reviewer: On the next line, I think you mean to say "is set to 0."

Reviewer: Fixed in c935bad.

>    When the majority of clients are IPv6-capable, we could set the default

Reviewer: ```suggestion
   When the majority of clients and relays are IPv6-capable, we could set the default
```

Reviewer: Fixed in c935bad.

   value of ClientPreferIPv6ORPort to 1 in order to take advantage of IPv6.

5. Relay Statistics

   Entry nodes could measure the following statistics from clients for both
   IPv4 and IPv6:

     * Number of successful connections 

     * Number of failed connections

     * Number of timed out connections

   Should we implement Section 5:

     * We could send this information to directory authorities

>      * We should consider the privacy implications of these statistics

Reviewer: ```suggestion
     * We should consider the privacy implications of these statistics, and how much noise to add to them
```

Reviewer: Fixed in c935bad.

>      * If submitted to dirauths, whether these should be public or not

Reviewer: Please delete this line: Tor relay statistics are all public.

Reviewer: Fixed in c935bad.

6. Initial Feasibility Testing

   We should test this proposal with the following scenarios:

    * Different combinations of values for the options ClientUseIPv4,
      ClientUseIPv6, and ClientPreferIPv6ORPort on IPv4-only, IPv6-only,
      and dual-stack connections

    * Dual-stack connections of different technologies, including
      high-bandwidth and low-latency (e.g. FTTH), moderate-bandwidth and
      moderate-latency (e.g. DSL, LTE), and high-latency and low-bandwidth
      (e.g. satellite, dial-up) to see if Prop306 is reliable and feasible

7. Minimum Viable Prop306 Product

   The mimumum viable product for Prop306 must include the following:

    * Implementation of the algorithm in Section 3.1 along with the changes

>       described in Sections 2.1 and 2.2

Reviewer: ```suggestion
      described in Sections 2.1 and 2.2 for client to entry node connections. (Single Onion services are optional, Bridge clients are out of scope.)
```

Reviewer: Fixed in 99c5b9b.

    * The Connection Success/Failure mechanism in Section 3.2 and Connection

>       Delay mechanism in Section 3.3 (the consensus parameter is optional)

Reviewer: ```suggestion
      Delay mechanism in Section 3.3 (the torrc option and consensus parameter are optional)
```

Reviewer: Fixed in 7a947b9.

    * A default setup capable of both IPv4 and IPv6 connections with the

>       options described in Section 4.1

Reviewer: ```suggestion
      options described in Section 4
```

Reviewer: Fixed in c935bad.

8. Optional Features

   Some features which are optional include:

> 

Reviewer: ```suggestion

    * Onion service protocol extend infos containing IPv4 and IPv6 addresses (used by Single Onion Services)
```

Reviewer: Fixed in 7a947b9.

>     * Consensus Parameter ConnectionAttemptDelay (Section 3.3) - We will need

Reviewer: ```suggestion
    * torrc options and Consensus Parameter ConnectionAttemptDelay (Section 3.3) - We will need
```

Reviewer: Fixed in 7a947b9.
